### PR TITLE
KAFKA-5117: Add password masking for kafka connect REST endpoint

### PR DESCRIPTION
**Pull request by qiao-meng-zefr**
*More detailed description of your change,
Mask all password type config parameter with "*********" instead of displaying the plain text in kafka connect REST endpoint.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by qiao-meng-zefr** [comment]
I'm new to kafka and not an expert on Java. So if anyone has better idea to achieve this, please let me know. I'm happy to make changes.

**Comment by qiao-meng-zefr** [comment]
@bharatviswa504 Sorry I can only find you since your name was involved in the ticket [KAFKA-5117](https://issues.apache.org/jira/browse/KAFKA-5117). Do you mind take a look at this PR or direct to anyone who can?

**Comment by qiao-meng-zefr** [comment]
@ewencp Could you take a look at this PR?

**Comment by qiao-meng-zefr** [comment]
@guozhangwang Do you mind take a look?

**Comment by guozhangwang** [comment]
@qiao-meng-zefr It seems from the JIRA someone has already started working on this ticket, have you coordinated with the other contributors for resolving this issue? cc @rhauch as well for reviewing.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: Connect code can't just load the class using the thread's classloader because of the plugin path feature. See below for suggestions.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: The `AbstractHerder` already has a `getConnector(String className)` that properly uses the plugin classloader, so it might be better to move this into the `AbstractHerder` class.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: We should use the `ConnectorConfig.CONNECTOR_CLASS_CONFIG` constant here, instead of creating a new literal value.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: Might be good to use a `LinkedHashMap` here to maintain the order of the map.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: How about looking up the `ConfigKey` once and then checking if it was found?
```
ConfigDef.ConfigKey configKey = definedConfigKeys.get(key);
if (configKey != null && ConfigDef.Type.Password.equals(configKey.type)) {
    ...
```
Also, note I reversed the comparison to be absolutely sure that there is not a NPE should `type` be null.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: Also, although this method is not public, it's probably useful to put this comment into a JavaDoc comment and to include information about the parameter. Can the `config` parameter be null? Will this method ever return a null value?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: Minor nit: why 9 asterisks, and not 8 or 10?

**Review by rhauch — CHANGES_REQUESTED** [review]
Thanks for the contribution, @qiao-meng-zefr! This looks good so far, but I do have a few requests below. Also, since `maskCredentials` is  a non-trivial, we really need unit tests to verify the logic, including with edge/error cases.

Also, as I mention on the JIRA, this changes in the Connect public API, and therefore we're going to need a KIP for this change to make sure that all considerations are taken into account. That won't change this PR, although before the PR can be merged we have to have an approved KIP. See https://cwiki.apache.org/confluence/display/KAFKA/Contributing+Code+Changes and https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals for details about that process.

**Comment by qiao-meng-zefr** [comment]
@guozhangwang Yeah I tried to find the person "Thomas Holmes" who was working on this in the People page in github, but couldn't find him. That jira issue seems old so I assumed no one is still working on it. Thanks for redirect this to rhauch.

@rhauch Thank you very much for the review I'll make change accordingly.

**Comment by Tang8330** [comment]
@qiao-meng-zefr @rhauch Any updates on this? I'm happy to help to get this change merged in.

**Comment by qiao-meng-zefr** [comment]
@Tang8330 Sorry I haven't got time to work on this in the past couple weeks. I will try to make those changes this week. After that I'll ping you for a review as well.

**Comment by qiao-meng-zefr** [comment]
Also there is a discussion thread. We might need to add a config parameter for user to opt in this for backwards compatible concerns.

**Comment by qiao-meng-zefr** [comment]
Also if you want to commit let me know, I can give you write access to my fork branch.

**Comment by Tang8330** [comment]
Sure, can you give me commit access @qiao-meng-zefr? I can try poking around.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (qiao-meng-zefr) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: It looks like after moving this method to the `AbstractHerder`, it's still not available to use in this class. Because the herder in this class is defined as a `Herder` not `AbstractHerder`. So it doesn't know the new method is available.

Should I add another method in the `Herder` interface `connectorMaskedConfig` which returns the masked config and I can implement that in the `AbstractHerder`? Any suggestions?

**Comment by qiao-meng-zefr** [comment]
@Tang8330 done!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: How about defining a new method on `Herder` that safely returns the `ConfigDef` given the name of a connector. This could then be implemented in `AbstractHerder` in terms of the existing methods that properly use the plugin classloaders, and could be called by this class to mask the configuration properties for both a connector and the tasks associated with a connector:

```
private Map<String, String> maskCredentials(String connectorName, Map<String, String> config) {
    ConfigDef configDef = herder.connectorConfigDef(connectorName);
    Map<String, ConfigDef.ConfigKey> definedConfigKeys = configDef.configKeys();
    ...
}
```

I'm not sure there's a way to get around adding a method to `Herder`. It is an interface, but it's not part of the public API and thus part of the implementation.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: BTW, the `Password` class is what's used within the AbstractConfig instances to hold onto a password literal so that when it is printed in the log the password is masked. It defines a `Password.HIDDEN` public static string that could be used here instead of the asterisks, and it would be more consistent with how they're hidden in the logs. WDYT?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (qiao-meng-zefr) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: That should work. I'll try that.

Is there a way in the AbstractHerder to get kafka connect config parameters? For backwards compatible we might need to add a parameter "enable.password.mask" for user to opt in. So I want to detect if this config is set to True in the AbstractHerder to decide if we should return the masked password or plain text.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (qiao-meng-zefr) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: Great. That's what I'm looking for. Done!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java
Comment (Tang8330) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: @rhauch I'm having a hard time exposing the `maskCredentials` method from `AbstractHerder` to be used by `ConnectorResource`...

Any suggestions? Should I also add `maskCredentials` into Herder interface?

**Comment by Tang8330** [comment]
FYI - CI builds are failing due to unused imports -- should be quarried away when we figure out how to expose methods from AbstractHerder to ConnectorResource.

**Comment by Tang8330** [comment]
Another request once this is accepted: are we able to back-port this into an older version of Kafka?
We use version 0.11.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: is `connName` correct here? `getConnector` takes a `connType` parameter. it looks like it already throws an exception if it can't find the class.

i want to make sure we get these correct because there has definitely been some confusion in the past and `maskCredentials` definitely looks like it is supposed to take the connector name whereas `getConnector` definitely takes the connector type (i.e. class name). You might need an extra hop through `AbstractHerder.config(String connName)` to get to the actual config, from which you can extract the connector class name.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: Can this be made polymorphic instead of actually checking the `PASSWORD` type here? e.g., could we simply always do `newConfig.put(key, entry.getValue().toString())` since `Password.toString()` always hides the value (in contrast to the `value()` method? I don't see a use atm, but keeping this generalizable seems potentially valuable.

wouldn't block on this, but it doesn't seem great to have to bake details of the config type into code in the `AbstractHerder` that otherwise doesn't care about it.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Herder.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Herder.java]: again, not a dealbreaker, but generally we try to avoid `get` prefixes on getters in Kafka code. there are a few exceptions, especially due to code originally being developed outside kafka, but best to avoid where possible.

**Review by ewencp — COMMENTED** [review]
I think there will be some more follow up with the config option, but I've left a couple of comments on the current PR.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (Tang8330) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: @ewencp Thanks so much for your suggestion! I see your point now. I am making the fixes as we speak.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Herder.java
Comment (Tang8330) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Herder.java]: Fair. Let's remove the get :)

**Comment by Tang8330** [comment]
Could I ping for a review? cc: @ewencp 
CI seems stuck: Jenkins stuck in reboot mode.

**Comment by qiao-meng-zefr** [comment]
Do we need to setup a config parameter for users to opt in? I think that's an valid argument. We don't want to break anything if people rely on this endpoint to get config.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: Can't the value of an entry be null?

Comment (Tang8330) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: Sorry, I have mistaken about the print null -- will add a check for this.

**Comment by Tang8330** [comment]
I am indifferent about the config parameters -- there's validity in keeping backwards compatibility.

However, exposing plain-text password may pose as a security risk. Also, this password is passed in, I'm not so sure that they would have a hard time retrieving the password.

**Comment by Tang8330** [comment]
@rhauch - I added a check to prevent null pointer exceptions now!

**Comment by Tang8330** [comment]
Currently running into 2 problems:
1. By simply ignoring `ConfigDef`, and going about the naive way of just iterating through key, values and doing `toString` is not going to work. `connectorInfo.config` returns a `Map<String, String>`, and the type "Password" is not being initialized.
2. Calling ConfigDef with the appropriate connType is erroring out: 

```
Failed to find any class that implements Connector and which name matches source, available connectors are: PluginDesc{klass=class com.google.pubsub.kafka.sink.CloudPubSubSinkConnector,
```

Still investigating...just wanted to give an update!

**Comment by Tang8330** [comment]
I have opened a new PR as this PR was getting big: https://github.com/apache/kafka/pull/4441

Please post in there! Thank you.

**Comment by qiao-meng-zefr** [comment]
A better solution is provided: https://cwiki.apache.org/confluence/display/KAFKA/KIP-297%3A+Externalizing+Secrets+for+Connect+Configurations

**Pull request closed without merging**
